_studio/mfx_lib/shared/src/libmfxsw_encode.cpp

```diff
@@ -169,57 +294,17 @@ mfxStatus MFXVideoENCODE_Query(mfxSession session, mfxVideoParam *in, mfxVideoPa
         else
 #endif
         {
-            switch (out->mfx.CodecId)
-            {
-#ifdef MFX_ENABLE_H264_VIDEO_ENCODE
-            case MFX_CODEC_AVC:
-#if defined (MFX_ENABLE_H264_VIDEO_ENCODE_HW)
-                if (!session->m_pENCODE.get())
-                    mfxRes = MFXHWVideoENCODEH264::Query(session->m_pCORE.get(), in, out);
-                else
-                    mfxRes = MFXHWVideoENCODEH264::Query(session->m_pCORE.get(), in, out, session->m_pENCODE.get());
-#else //MFX_VA
-                mfxRes = MFXVideoENCODEH264::Query(in, out);
-#endif //MFX_VA
-                break;
-#endif
-
-#ifdef MFX_ENABLE_MPEG2_VIDEO_ENCODE
-            case MFX_CODEC_MPEG2:
-                mfxRes = MFXVideoENCODEMPEG2_HW::Query(session->m_pCORE.get(), in, out);
-                break;
-#endif
-
-#if defined(MFX_ENABLE_MJPEG_VIDEO_ENCODE)
-            case MFX_CODEC_JPEG:
-                mfxRes = MFXVideoENCODEMJPEG_HW::Query(session->m_pCORE.get(), in, out);
-                break;
-#endif // MFX_ENABLE_MJPEG_VIDEO_ENCODE
-
-#if defined(MFX_ENABLE_H265_VIDEO_ENCODE)
-            case MFX_CODEC_HEVC:
-            {
 #if defined (MFX_ENABLE_HEVC_VIDEO_FEI_ENCODE)
-                bool * feiEnabled = (bool*)session->m_pCORE->QueryCoreInterface(MFXIFEIEnabled_GUID);//required to check FEI plugin registration.
-                MFX_CHECK_NULL_PTR1(feiEnabled);
-                if(*feiEnabled)
-                    mfxRes = MfxHwH265FeiEncode::H265FeiEncode_HW::Query(session->m_pCORE.get(), in, out);
-                else
+            // required to check FEI plugin registration
+            bool *feiEnabled = (bool*)session->m_pCORE->QueryCoreInterface(MFXIFEIEnabled_GUID);
+            MFX_CHECK_NULL_PTR1(feiEnabled);
+            const bool fei = *feiEnabled;
+#else
+            const bool fei = false;
 #endif
-                    mfxRes = MfxHwH265Encode::MFXVideoENCODEH265_HW::Query(session->m_pCORE.get(), in, out);
-                break;
-            }
-#endif // MFX_ENABLE_H265_VIDEO_ENCODE
-
-#if defined(MFX_ENABLE_VP9_VIDEO_ENCODE)
-            case MFX_CODEC_VP9:
-                mfxRes = MfxHwVP9Encode::MFXVideoENCODEVP9_HW::Query(&session->m_coreInt, in, out);
-                break;
-#endif //MFX_ENABLE_VP9_VIDEO_ENCODE
-
-            default:
-                mfxRes = MFX_ERR_UNSUPPORTED;
-            } // switch (out->mfx.CodecId)
+            auto handlers = codecId2Handlers.find(CodecKey(out->mfx.CodecId, fei));
```

Comment:
I think you can try to dispose of `#ifdef MFX_ENABLE_USER_ENCODE` section in the same way you did with FEI. I.e. add one more entry for the key:
```
struct CodecKey {
    const mfxU32 codecId;
    const bool   fei;
    const bool plugin;
}
```

And define
```
static const CodecId2Handlers codecId2Handlers =
{
 #ifdef MFX_ENABLE_USER_ENCODE
   {
       { .plugin=true, },
       { [](mfxSession s){ s->m_plgEnc->Query(session->m_pCORE.get(), in, out); } }
   }
#endif
}

```

You **must make sure** however that plugin entry is always the first to be found in a map, i.e. if `codecId2Handlers.find(CodecKey(out->mfx.CodecId, fei, isPlugin=true))`  should fall to plugin, not to codec. (user may register its own implementation of AVC as a plugin and it should have a preference).

@mgonchar : how do you feel about this idea?

Reply:
I see that both `query` and `queryIOSurf` for plugin are codec-agnostic. So, can we use single universal handles for every plugin? And map it to some forever-unused codec ID. The code is unified, and plugin preference is provided automatically. One is added as a separate commit.

Reply:
Yes, you should have single entry for the plugins like {.isplugin=true, .isfei=false, .codecid=0 }